controller.go

```diff
@@ -562,177 +535,89 @@ func (c *Controller) syncHandler(key string) error {
 				Additional: map[string]string{},
 			})
 
-			if err != nil {
-				return err
-			}
-			id = res.ID
+		} else if c.infraConfig.Provider == "linode" {
+			provisioner, _ := provision.NewLinodeProvisioner(c.infraConfig.GetAccessKey())
 
+			res, err = provisioner.Provision(provision.BasicHost{
+				Name:       tunnel.Name,
+				OS:         "linode/ubuntu16.04lts", // https://api.linode.com/v4/images
+				Plan:       "g6-nanode-1",           // https://api.linode.com/v4/linode/types
+				Region:     c.infraConfig.Region,
+				UserData:   userData,
+				Additional: map[string]string{},
+			})
+
+		} else {
+			return fmt.Errorf("unsupported provider: %s", c.infraConfig.Provider)
+		}
+
+		// If an error occurs during Update, we'll requeue the item so we can
+		// attempt processing again later. THis could have been caused by a
+		// temporary network failure, or any other transient reason.
+		if err != nil {
+			return err
 		}
 
 		log.Printf("Provisioning call took: %fs\n", time.Since(start).Seconds())
 
+		err = c.updateTunnelProvisioningStatus(tunnel, "provisioning", res.ID, "")
 		if err != nil {
-			err = c.updateTunnelProvisioningStatus(tunnel, "error", "", "")
-			if err != nil {
-				return err
-			}
-		} else {
-			err = c.updateTunnelProvisioningStatus(tunnel, "provisioning", id, "")
-			if err != nil {
-				return fmt.Errorf("tunnel %s (%s) update error: %s", tunnel.Name, "provisioning", err)
-			}
+			return fmt.Errorf("tunnel %s (%s) update error: %s", tunnel.Name, "provisioning", err)
 		}
 
 		break
 
 	case "provisioning":
-
+		var provisioner provision.Provisioner
 		if c.infraConfig.Provider == "packet" {
-
```

Comment:
Can you explain the refactoring here?

Reply:
When I was editing that part of code.  I find it a bit hard to read as it has duplicate lines for each provider, making the whole switch block to be several screen scrolls.

So I figured that I should refactor a bit here. It would also make it easier for future new providers.

I understand usually we shouldn't refactor when adding new feature, but this refactor is not really a big refactor, so I figured it may be fine.

I'm okay with separating this refactoring as another PR if needed.

Reply:
Would you mind "only doing one thing" (tm) to help me preview and verify the code change?

I'm open to refactoring the codeblock in a follow-up PR - I think @aledbf has some similar ideas to this too.

Reply:
I have reverted the switch block refactoring, will submit another PR for that after this is merged.